internal/connection_manager.go

```diff
@@ -138,23 +154,31 @@ func (cm *connectionManager) ConnectionCount() int32 {
 	return int32(len(cm.connections))
 }
 
-func (cm *connectionManager) openNewConnection(address core.Address, resp chan *Connection, asOwner bool) error {
+func (cm *connectionManager) getOrConnectInternal(address core.Address, asOwner bool) (*Connection, error) {
+	cm.lock.Lock()
```

Comment:
Instead of using write lock and checking if connection exists, would it be better to first check if connection exists with `getConnection` which is guarded under read lock instead of `getConnectionInternal` which is not, and then if it doesnt exist require write locks and connect.

Reply:
I guess that would result in opening multiple connections though so nevermind

Reply:
I have put the reading connection under readlock where this method is called. This is the optimization part. 
We need to check the `getConnectionInternal` under writeLock again to make sure not to open multiple connections.